controllers/component/component_adapter.go

```diff
@@ -0,0 +1,161 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package component
+
+import (
+	"context"
+
+	applicationapiv1alpha1 "github.com/redhat-appstudio/application-api/api/v1alpha1"
+	"github.com/redhat-appstudio/integration-service/gitops"
+	h "github.com/redhat-appstudio/integration-service/helpers"
+	"github.com/redhat-appstudio/integration-service/loader"
+	"github.com/redhat-appstudio/integration-service/metrics"
+	"github.com/redhat-appstudio/operator-toolkit/controller"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Adapter holds the objects needed to reconcile a integration PipelineRun.
+type Adapter struct {
+	component   *applicationapiv1alpha1.Component
+	application *applicationapiv1alpha1.Application
+	loader      loader.ObjectLoader
+	logger      h.IntegrationLogger
+	client      client.Client
+	context     context.Context
+}
+
+// NewAdapter creates and returns an Adapter instance.
+func NewAdapter(component *applicationapiv1alpha1.Component, application *applicationapiv1alpha1.Application, logger h.IntegrationLogger, loader loader.ObjectLoader, client client.Client,
+	context context.Context) *Adapter {
+	return &Adapter{
+		component:   component,
+		application: application,
+		logger:      logger,
+		loader:      loader,
+		client:      client,
+		context:     context,
+	}
+}
+
+// EnsureComponentIsCleanedUp is an operation that will ensure components
+// marked for deletion have a snapshot created without said component
+func (a *Adapter) EnsureComponentIsCleanedUp() (controller.OperationResult, error) {
+	if !hasComponentBeenDeleted(a.component) {
+		return controller.ContinueProcessing()
+	}
+
+	applicationComponents, err := a.loader.GetAllApplicationComponents(a.client, a.context, a.application)
+	if err != nil {
+		a.logger.Error(err, "Failed to load application components")
+		return controller.RequeueWithError(err)
+	}
+
+	var snapshotComponents []applicationapiv1alpha1.SnapshotComponent
+
+	for _, individualComponent := range *applicationComponents {
+		component := individualComponent
+		if a.component.Name != component.Name {
+			containerImage := component.Spec.ContainerImage
+			componentSource := gitops.GetComponentSourceFromComponent(&component)
+			snapshotComponents = append(snapshotComponents, applicationapiv1alpha1.SnapshotComponent{
+				Name:           component.Name,
+				ContainerImage: containerImage,
+				Source:         *componentSource,
+			})
+		}
+	}
+
+	err = a.failStaleSnapshots()
```

Comment:
I've been giving this some thought and I think that marking all Snapshots that contain the deleted component as failed testing would introduce a number of problems, while not providing a large benefit.

The downsides to doing this are:

- Updating the status of all (or majority of) previous Snapshots would be very resource intensive (there could be potentially up to a thousand of old Snapshots in the user's workspace).
- The update events would/could trigger reconciliations in other services, compounding the resource issues
- Marking stale Snapshots as failed would send a wrong message to the end-users since it's not the tests that failed, but the validity of the Snapshot itself

The benefit of marking them as failed is also lessened in light of newer changes (such as #340) that prevent old Snapshots from being promoted without being tested again - which would immediately reveal that their component composition is out-of-date anyways.

Reply:
ack this makes sense, I hadn't considered some of those potential issues that could arrive from iterating over all that historical data. I removed the code involved in failing snapshots, but @dirgim can you confirm that would extend to snapshots currently in testing as well?

Reply:
That's a good question - when a Snapshot finishes testing, the integrationpipeline adapter first checks if the components in the Snapshot match the current state of the global candidate list (GCL) ([here](https://github.com/redhat-appstudio/integration-service/blob/92c8f944f0104ba4d687b89bd458008cd341bcb9/controllers/integrationpipeline/integrationpipeline_adapter.go#L117-L120)).
Since the Snapshot that just ended testing would be missing a component, it should be marked as invalid and the integrationpipeline adapter should check if there's already a fresher Snapshot that matches the GCL. In our scenario here, the component_adapter would have already created that fresh Snapshot so the logic would stop there.

Reply:
Resolving this convo so it doesn't block the PR from going forward.